src/rhash/hash.c

```diff
@@ -868,6 +880,105 @@ static int rc_hash_pcfx_cd(char hash[33], const char* path)
   return rc_hash_finalize(&md5, hash);
 }
 
+static int rc_hash_dreamcast(char hash[33], const char* path)
+{
+  uint8_t buffer[2048];
+  void* track_handle;
+  void* last_track_handle;
+  char exe_file[64] = "";
+  unsigned size;
+  size_t num_read = 0;
+  uint32_t sector;
+  uint32_t last_track_lba;
+  int result = 0;
+  md5_state_t md5;
+  int i = 0;
+
+  /* track 03 is the data track that contains the directory record and IP.BIN */
+  track_handle = rc_cd_open_track(path, 3);
+  if (!track_handle)
+    return rc_hash_error("Could not open track");
+
+  /* first 256 bytes from first sector should have IP.BIN structure that stores game meta information 
+    https://mc.pp.se/dc/ip.bin.html */
+  rc_cd_read_sector(track_handle, 0, buffer, sizeof(buffer));
+
+  if (memcmp(&buffer[0], "SEGA SEGAKATANA ", 16) != 0) 
+  {
+    rc_cd_close_track(track_handle);
+    return rc_hash_error("Not a Dreamcast CD");
+  }
+
+  if (verbose_message_callback)
+  {
+    char message[256];
+    snprintf(message, sizeof(message), "Hashing meta information:\nSoftware Name = %.127s\nProduct Number = %.9s\nProduct Version = %.5s\n",
+                                        &buffer[0x80], &buffer[0x40], &buffer[0x4A]);
+    verbose_message_callback(message);
+  }
+
+  md5_init(&md5);
+  md5_append(&md5, (md5_byte_t*)buffer, 256);
+
+  /* remove whitespace from bootfile*/
+  for (i = 0; i < 16; i++)
+    if (!isspace(buffer[96 + i]))
+      exe_file[i] = buffer[96 + i];
+
+  /* sometimes boot file isn't present on meta information.
+     nothing can be done, as even the core doesn't run the game in this case. */
+  if (!strlen(exe_file))
+  {
+    rc_cd_close_track(track_handle);
+    return rc_hash_error("Boot executable not specified on IP.BIN");
+  }
+  
+  /* last track contains the boot executable */
+  last_track_handle = rc_cd_open_track(path, 0);
+  last_track_lba = rc_cd_get_lba(last_track_handle);
+
+  /* its offset being 45000 means it is track 03.
+     if so, start finding executable from here. */
+  if (last_track_lba == 45000)
+    sector = rc_cd_find_file_sector(last_track_handle, exe_file, &size, last_track_lba);
+  else
+    sector = rc_cd_find_file_sector(track_handle, exe_file, &size, rc_cd_get_lba(track_handle));
+  
+  rc_cd_close_track(track_handle);
+
+  /* subtract it by its LBA to get the true sector location on the file */
+  sector -= last_track_lba;
+
+  if (!sector)
+    rc_hash_error("Could not locate boot executable");
```

Comment:
this check needs to occur before subtracting `last_track_lba` and should exit the function, as `num_read` and `buffer` will be invalid going into the loop.